tools/fix_mailmap.py

```diff
@@ -0,0 +1,573 @@
+
+    1. Warn on non-standard lines
+    2. Remove name-matching
+    3. Drop exact duplicate lines
+    4. Warn for partial duplicate lines
+    5. Drop lines that don't match a commit
+    6. Drop lines that match a commit but cause no changes
+    7. Alphabetise the body
+    8. Warn about multiple canonical names associated with same email (we assume project-unique names)
+    9. Warn about multiple canonical emails associated with same name (we assume project-unique emails)
+    """
+    line_re = re.compile(r"^([^<]+)\s+<([^>]*)>(?:(\s+[^<]+)?\s+<([^>]*)>)?$")
+    output_lines = []  # Transformed lines with fixes applied
+    email_mapping: dict[str, tuple] = {}  # Mapping from match-email to (canon-name, email, index)
+    canonical_emails: dict[str, set[str]] = {}  # Mapping from canonical-email to {canonical-name}
+    canonical_names: dict[str, set[str]] = {}  # Mapping from canonical-name to {canonical-email}
+
+    for index, raw_line in enumerate(mailmap_lines):
+        line, *comment = raw_line.split("#")
+        line = line.strip()
+        match = line_re.fullmatch(line)
+        if not match:
+            print_red(
+                f"Found non-canonical mailmap line {mailmap_line(index)} "
+                f"Please fix manually:\n    {raw_line}"
+            )
+            output_lines.append(raw_line)
+            continue
+        canonical_name, canonical_email, name2, email2 = match.group(1, 2, 3, 4)
+        match_email = canonical_email if email2 is None else email2
+        match_email = match_email.lower()  # git finds matches on lowercase of name, email
+
+        # We could extend this tool to handle name matching if there's ever a case where it's needed.
+        # But currently our history has no cases where multiple contributors share a commit email.
+        if match_email.lower() not in raw_authors:
+            print_yellow(
+                f"Dropped the mailmap line {mailmap_line(index)} for <{match_email}> "
+                "that doesn't match any author of a commit"
+            )
+            continue
+        if match_email == canonical_email and raw_authors[match_email] == {
+            (canonical_name, canonical_email)
+        }:
+            print_yellow(
+                f"Dropped the mailmap line {mailmap_line(index)} for <{match_email}> "
+                "that matches but doesn't cause any change",
+            )
+            continue
+
+        if name2 is not None:
+            print_yellow(
+                f"Dropped mailmap name-match {mailmap_line(index)}:\n    {raw_line}",
+            )
+            line = f"{canonical_name} <{canonical_email}> <{email2}>"
+        # if email2 == canonical_email:
+        #     print_yellow(
+        #         f"Dropped mailmap superfluous second email <{email2}> that is equal to the first "
+        #         f"{mailmap_line(index)}"
+        #     )
+        #     line = f"{canonical_name} <{canonical_email}>"
+
```

Comment:
Did you want to clean this out?

Reply:
It tries to remove the cases like you pointed out for Ismael's entry only ones where the second actually email isn't needed. But it is sometimes over-zealous and I couldn't think of an easy way to get it to only trigger in the needed cases. I'll add a TODO comment to clarify and maybe someone else has an idea